app/controllers/sitemap_controller.rb

```diff
@@ -0,0 +1,47 @@
+class SitemapController < ApplicationController
+
+  def sitemap
+    #if false
+      #redirect_to "/" # The CDN then fetches the sitemap.xml from the sitemap/generate URL which points to the `generate` action
+    #else
+    render_site_map
+    #end
+  end
+
+  def generate
+    render_site_map
+  end
+
+  private
+
+    def render_site_map
+      respond_to do |format|
+
+        data = Rails.cache.fetch(@current_community, expires_in: 24.hours) do
+          adapter = SitemapGenerator::NeverWriteAdapter.new
+
+          if @current_community.use_domain 
+            default_host = "https://"+@current_community.domain
+          else 
+            default_host = "https://"+APP_CONFIG.domain
+          end
+
+          community_id = @current_community.id
+
+          SitemapGenerator::Sitemap.create(
+                :default_host => default_host,
+                :verbose => false,
+                :adapter => adapter) do
+                  Listing.where(deleted: false, open: true, community_id: community_id).find_each do |listing|
```

Comment:
This needs to be tuned a bit. It's now over-fetching listings because it doesn't take into account the `valid_until` date. There's a `currently_open` method in `Listing` model that is useful for this purpose.

I deployed this branch to our testing server and tested the performance with one of our test marketplaces which has tens of thousands listings.

The `GET /sitemap.xml.gz` was completed in ~5000ms (with empty cache), which is way too much. Something like ~300ms would be sufficient.

The reason for this was the over-fetching. `Listing.where(deleted: false, open: true, community_id: community_id)` returned ~9000 listings.

Then I deployed a modified version (using `Listing.currently_open.where(community_id: community_id)` instead). Now ~900 listings were returned and the rendering took something like ~500ms. Not exactly a fast request, but good enough in my opinion, because the request is cached heavily.

In addition, [pluck](http://apidock.com/rails/ActiveRecord/Calculations/pluck) might make the data fetching a bit faster. Because you only need `id` and `updated_at`, `.pluck(:id, :updated_at)` can accomplish the same job.